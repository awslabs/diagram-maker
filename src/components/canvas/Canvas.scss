@import './patterns/DotPattern.scss';
@import './patterns/GridPattern.scss';

.dm-canvas {
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: #f2f3f3;
  pointer-events: none;
  user-select: none;
  z-index: 1;

  > svg {
    position: absolute;
    width: 100%;
    height: 100%;
  }
}
