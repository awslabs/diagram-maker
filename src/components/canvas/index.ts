export { default as Canvas, CanvasProps } from './Canvas';
