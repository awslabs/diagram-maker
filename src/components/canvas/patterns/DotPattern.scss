.dm-dots {
  shape-rendering: geometricPrecision;

  pattern {
    path {
      fill: #d5dbdb;
    }
  }
}
