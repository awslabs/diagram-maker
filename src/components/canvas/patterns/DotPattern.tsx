import * as Preact from 'preact';

export interface DotPatternProps {
  cellSize: number;
  radius: number;
}

const getQuarterCirclePathData = (cx: number, cy: number, r: number): string => {
  return `M${cx},${cy} m${-r},0 a${r},${r} 0 1,1 ${r * 2},0 a${r},${r} 0 1,1 ${-r * 2},0`;
};

const DotPattern = (props: DotPatternProps): JSX.Element => {
  const { cellSize, radius } = props;
  const pathTL = getQuarterCirclePathData(0, 0, radius);
  const pathTR = getQuarterCirclePathData(cellSize, 0, radius);
  const pathBR = getQuarterCirclePathData(cellSize, cellSize, radius);
  const pathBL = getQuarterCirclePathData(0, cellSize, radius);

  return (
    <svg className="dm-dots" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <pattern id="dm-dot-pattern" width={cellSize} height={cellSize} patternUnits="userSpaceOnUse">
          <path d={pathTL} />
          <path d={pathTR} />
          <path d={pathBR} />
          <path d={pathBL} />
        </pattern>
      </defs>
      <rect fill="url(#dm-dot-pattern)" width="100%" height="100%" />
    </svg>
  );
};

export default DotPattern;
