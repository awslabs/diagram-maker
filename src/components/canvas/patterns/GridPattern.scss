.dm-grid {
  shape-rendering: geometricPrecision;

  pattern {
    path {
      fill: none;
      stroke: #d5dbdb;
    }
  }
}
