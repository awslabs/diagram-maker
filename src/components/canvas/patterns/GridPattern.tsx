import * as Preact from 'preact';

export interface GridPatternProps {
  cellSize: number;
  patternSize: number;
}

const GridPattern = (props: GridPatternProps): JSX.Element => {
  const { cellSize, patternSize } = props;

  return (
    <svg className="dm-grid" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <pattern id="dm-grid-smallpattern" width={cellSize} height={cellSize} patternUnits="userSpaceOnUse">
          <path d={`M ${cellSize} 0 L 0 0 0 ${cellSize}`} stroke-width="0.5" />
        </pattern>
        <pattern id="dm-grid-pattern" width={patternSize} height={patternSize} patternUnits="userSpaceOnUse">
          <rect width={patternSize} height={patternSize} fill="url(#dm-grid-smallpattern)" />
          <path d={`M ${patternSize} 0 L 0 0 0 ${patternSize}`} stroke-width="1" />
        </pattern>
      </defs>
      <rect fill="url(#dm-grid-pattern)" width="100%" height="100%" />
    </svg>
  );
};

export default GridPattern;
