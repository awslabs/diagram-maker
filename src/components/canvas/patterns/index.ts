export { default as DotPattern, DotPatternProps } from './DotPattern';
export { default as GridPattern, GridPatternProps } from './GridPattern';
