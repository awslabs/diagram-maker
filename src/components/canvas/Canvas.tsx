import * as classnames from 'classnames';
import * as Preact from 'preact';

import './Canvas.scss';

export interface CanvasProps {
  className?: string | string[];
}

const Canvas = (props: Preact.RenderableProps<CanvasProps>): JSX.Element => {
  const className = classnames('dm-canvas', props.className);

  return (
    <div className={className}>
        {props.children}
    </div>
  );
};

export default Canvas;
