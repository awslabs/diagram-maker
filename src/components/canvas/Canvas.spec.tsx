import * as Preact from 'preact';
import { shallow } from 'preact-render-spy';

import { Canvas } from 'diagramMaker/components/canvas';
import { DotPattern, GridPattern } from 'diagramMaker/components/canvas/patterns';

describe('Canvas', () => {
  it('renders canvas given dot pattern', () => {
    const dotPattern = <DotPattern cellSize={10} radius={1} />;

    const canvas = shallow(
            <Canvas>
              {dotPattern}
            </Canvas>
        );

    expect(canvas).toMatchSnapshot();
  });

  it('renders canvas given grid pattern', () => {
    const gridPattern = <GridPattern cellSize={10} patternSize={100}/>;

    const canvas = shallow(
        <Canvas>
          {gridPattern}
        </Canvas>
    );

    expect(canvas).toMatchSnapshot();
  });
});
