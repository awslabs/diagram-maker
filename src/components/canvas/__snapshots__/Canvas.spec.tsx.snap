// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas renders canvas given dot pattern 1`] = `
preact-render-spy (1 nodes)
-------
<div class="dm-canvas">
  <DotPattern
    cellSize={10}
    radius={1}
  >
  </DotPattern>
</div>

`;

exports[`Canvas renders canvas given grid pattern 1`] = `
preact-render-spy (1 nodes)
-------
<div class="dm-canvas">
  <GridPattern
    cellSize={10}
    patternSize={100}
  >
  </GridPattern>
</div>

`;
